File: client/android/app/src/main/java/com/stonefive/chalkak/feature/settings/SettingsScreen.kt
package com.stonefive.chalkak.feature.settings

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.stonefive.chalkak.R
import com.stonefive.chalkak.core.designsystem.component.bottombar.ChalkakBottomBar
import com.stonefive.chalkak.core.designsystem.component.bottombar.ChalkakBottomBarItem
import com.stonefive.chalkak.core.designsystem.component.dialog.ChalkakConfirmDialog
import com.stonefive.chalkak.core.designsystem.component.dialog.ChalkakConfirmDialogStyle
import com.stonefive.chalkak.core.designsystem.theme.ChalkakTheme
import com.stonefive.chalkak.core.ui.UiMessageEffect
import com.stonefive.chalkak.feature.settings.component.SettingsAccountCard
import com.stonefive.chalkak.feature.settings.component.SettingsInformationCard
import com.stonefive.chalkak.feature.settings.component.SettingsLoginButton
import com.stonefive.chalkak.feature.settings.component.SettingsSectionLabel
import com.stonefive.chalkak.feature.settings.component.SettingsSignatureCard

@Composable
fun SettingsRoute(
    onNavigateToSignature: () -> Unit,
    onOpenPrivacyPolicy: () -> Unit,
    onOpenTerms: () -> Unit,
    onNavigateToBottomBar: (ChalkakBottomBarItem) -> Unit,
    onOpenPhotoUpload: () -> Unit,
    signatureUpdateUrl: String? = null,
    viewModel: SettingsViewModel = viewModel(factory = SettingsViewModel.Factory),
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    UiMessageEffect(uiState.pendingMessage, viewModel::onMessageShown)

    LaunchedEffect(signatureUpdateUrl) {
        signatureUpdateUrl?.let(viewModel::applySignatureUpdate)
    }

    SettingsScreen(
        uiState = uiState,
        onLoginClick = viewModel::startLogin,
        onChangeSignatureClick = onNavigateToSignature,
        onPrivacyPolicyClick = onOpenPrivacyPolicy,
        onTermsClick = onOpenTerms,
        onLogoutClick = viewModel::showLogoutDialog,
        onWithdrawClick = viewModel::showWithdrawDialog,
        onAccountDialogConfirm = viewModel::confirmAccountAction,
        onAccountDialogDismiss = viewModel::dismissAccountDialog,
        onNavigateToBottomBar = onNavigateToBottomBar,
        onAddClick = onOpenPhotoUpload,
    )
}

@Composable
fun SettingsScreen(
    uiState: SettingsUiState,
    onLoginClick: () -> Unit,
    onChangeSignatureClick: () -> Unit,
    onPrivacyPolicyClick: () -> Unit,
    onTermsClick: () -> Unit,
    onLogoutClick: () -> Unit,
    onWithdrawClick: () -> Unit,
    onAccountDialogConfirm: () -> Unit,
    onAccountDialogDismiss: () -> Unit,
    onNavigateToBottomBar: (ChalkakBottomBarItem) -> Unit,
    onAddClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    uiState.accountDialog?.let { accountDialog ->
        ChalkakConfirmDialog(
            title = accountDialog.title,
            message = accountDialog.message,
            confirmText = accountDialog.confirmText,
            onConfirm = onAccountDialogConfirm,
            onDismiss = onAccountDialogDismiss,
            modifier = Modifier.width(AccountDialogWidth),
            confirmStyle = accountDialog.confirmStyle,
        )
    }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        containerColor = ChalkakTheme.colors.background,
        bottomBar = {
            ChalkakBottomBar(
                selectedItem = ChalkakBottomBarItem.SETTINGS,
                onItemSelected = onNavigateToBottomBar,
                onAddClick = onAddClick,
                modifier = Modifier.fillMaxWidth(),
            )
        },
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .padding(horizontal = ChalkakTheme.spacing.screenHorizontal)
                .verticalScroll(rememberScrollState()),
        ) {
            Spacer(modifier = Modifier.height(48.dp))

            SettingsSectionLabel(text = "앱 설정")

            Spacer(modifier = Modifier.height(16.dp))

            when {
                uiState.isLoading -> Spacer(modifier = Modifier.height(56.dp))

                uiState.isLoggedIn -> SettingsSignatureCard(
                    signatureUrl = uiState.signatureUrl,
                    onChangeClick = onChangeSignatureClick,
                    modifier = Modifier.fillMaxWidth(),
                )

                else -> SettingsLoginButton(
                    onClick = onLoginClick,
                    modifier = Modifier.fillMaxWidth(),
                )
            }

            Spacer(modifier = Modifier.height(36.dp))

            SettingsSectionLabel(text = "정보 및 약관")

            Spacer(modifier = Modifier.height(16.dp))

            SettingsInformationCard(
                versionName = uiState.versionName,
                onPrivacyPolicyClick = onPrivacyPolicyClick,
                onTermsClick = onTermsClick,
                modifier = Modifier.fillMaxWidth(),
            )

            if (uiState.isLoggedIn) {
                Spacer(modifier = Modifier.height(40.dp))
                SettingsAccountCard(
                    onLogoutClick = onLogoutClick,
                    onWithdrawClick = onWithdrawClick,
                    enabled = !uiState.isAccountActionInProgress,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

@Preview(showBackground = true, widthDp = 390, heightDp = 844)
@Composable
private fun AuthenticatedSettingsScreenPreview() {
    ChalkakTheme {
        SettingsScreenPreview(
            uiState = SettingsUiState(
                isLoggedIn = true,
                signatureUrl = "android.resource://com.stonefive.chalkak/${R.drawable.preview_signature}",
                versionName = "1.0",
            ),
        )
    }
}

@Preview(showBackground = true, widthDp = 390, heightDp = 844)
@Composable
private fun GuestSettingsScreenPreview() {
    ChalkakTheme {
        SettingsScreenPreview(
            uiState = SettingsUiState(
                versionName = "1.0",
            ),
        )
    }
}

@Composable
private fun SettingsScreenPreview(uiState: SettingsUiState) {
    SettingsScreen(
        uiState = uiState,
        onLoginClick = {},
        onChangeSignatureClick = {},
        onPrivacyPolicyClick = {},
        onTermsClick = {},
        onLogoutClick = {},
        onWithdrawClick = {},
        onAccountDialogConfirm = {},
        onAccountDialogDismiss = {},
        onNavigateToBottomBar = {},
        onAddClick = {},
    )
}

private val SettingsAccountDialog.title: String
    get() = when (this) {
        SettingsAccountDialog.LOGOUT -> "로그아웃"
        SettingsAccountDialog.WITHDRAW -> "회원탈퇴"
    }

private val SettingsAccountDialog.message: String
    get() = when (this) {
        SettingsAccountDialog.LOGOUT -> "정말 로그아웃 하시겠습니까?"
        SettingsAccountDialog.WITHDRAW -> "정말 회원탈퇴 하시겠습니까?"
    }

private val SettingsAccountDialog.confirmText: String
    get() = title

private val SettingsAccountDialog.confirmStyle: ChalkakConfirmDialogStyle
    get() = when (this) {
        SettingsAccountDialog.LOGOUT -> ChalkakConfirmDialogStyle.PRIMARY
        SettingsAccountDialog.WITHDRAW -> ChalkakConfirmDialogStyle.DESTRUCTIVE
    }

private val AccountDialogWidth = 317.dp

File: client/iOS/chalkak/chalkak/Features/Settings/SettingsScreen.swift
import Foundation
import SwiftUI

enum SettingsAccountDialog: Equatable {
    case logout
    case withdraw

    var title: String {
        switch self {
        case .logout: "로그아웃"
        case .withdraw: "회원탈퇴"
        }
    }

    var message: String {
        switch self {
        case .logout: "정말 로그아웃 하시겠습니까?"
        case .withdraw: "정말 회원탈퇴 하시겠습니까?"
        }
    }

    var style: ChalkakConfirmDialogStyle {
        switch self {
        case .logout: .primary
        case .withdraw: .destructive
        }
    }
}

struct SettingsScreen: View {
    @Environment(\.chalkakTheme) private var theme

    @Bindable var viewModel: SettingsViewModel
    var onLogin: () -> Void = {}
    var onPrivacyPolicy: () -> Void = {}
    var onTerms: () -> Void = {}
    var onSignedOut: () -> Void = {}
    var onNavigateToBottomBar: (ChalkakBottomBarItem) -> Void = { _ in }
    var onOpenPhotoUpload: () -> Void = {}

    @State private var message: String?
    @State private var messageDismissTask: Task<Void, Never>?
    @State private var isSignatureChangePresented = false

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                SettingsSectionLabel(text: "앱 설정")

                Group {
                    if viewModel.viewState.isLoading {
                        Color.clear.frame(height: Metrics.loadingHeight)
                    } else if viewModel.viewState.isLoggedIn {
                        SettingsSignatureCard(
                            signatureSource: viewModel.viewState.signatureURL.map(ChalkakImageSource.remote),
                            onChange: { isSignatureChangePresented = true }
                        )
                    } else {
                        SettingsLoginButton(onClick: onLogin)
                    }
                }
                .padding(.top, theme.spacing.lg)

                SettingsSectionLabel(text: "정보 및 약관")
                    .padding(.top, Metrics.appToInformationSpacing)

                SettingsInformationCard(
                    version: viewModel.viewState.version,
                    onPrivacyPolicy: onPrivacyPolicy,
                    onTerms: onTerms
                )
                .padding(.top, theme.spacing.lg)

                if viewModel.viewState.isLoggedIn {
                    SettingsAccountCard(
                        onLogout: viewModel.showLogoutDialog,
                        onWithdraw: viewModel.showWithdrawDialog,
                        isEnabled: !viewModel.viewState.isAccountActionInProgress
                    )
                    .padding(.top, theme.spacing.xxl)
                }
            }
            .padding(.horizontal, theme.spacing.screenHorizontal)
            .padding(.top, Metrics.topPadding)
            .padding(.bottom, theme.spacing.xxl)
        }
        .scrollIndicators(.hidden)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(theme.colors.background)
        .safeAreaInset(edge: .bottom, spacing: 0) {
            ChalkakBottomBar(
                selectedItem: .settings,
                onSelect: onNavigateToBottomBar,
                onAdd: onOpenPhotoUpload
            )
        }
        .overlay {
            if let accountDialog = viewModel.viewState.accountDialog {
                ChalkakConfirmDialog(
                    title: accountDialog.title,
                    message: accountDialog.message,
                    confirmText: accountDialog.title,
                    confirmStyle: accountDialog.style,
                    onConfirm: { Task { await viewModel.confirmAccountAction() } },
                    onDismiss: viewModel.dismissAccountDialog
                )
                .transition(.opacity.combined(with: .scale(scale: 0.98)))
            }
        }
        .overlay(alignment: .bottom) { toast }
        .animation(.easeOut(duration: 0.16), value: viewModel.viewState.accountDialog)
        .task {
            await viewModel.load()
        }
        .onChange(of: viewModel.event) { _, event in
            handle(event)
        }
        .onDisappear {
            messageDismissTask?.cancel()
        }
        .fullScreenCover(isPresented: $isSignatureChangePresented) {
            SignatureChangeFlow(onUpdate: viewModel.updateSignature)
        }
    }

    @ViewBuilder
    private var toast: some View {
        if let message {
            Text(message)
                .font(theme.typography.subheadline)
                .foregroundStyle(theme.colors.onActionPrimary)
                .padding(.horizontal, theme.spacing.lg)
                .padding(.vertical, theme.spacing.md)
                .background(theme.colors.actionPrimary, in: Capsule())
                .padding(.bottom, Metrics.toastBottomPadding)
                .transition(.move(edge: .bottom).combined(with: .opacity))
                .accessibilityLabel(message)
        }
    }

    private func handle(_ event: SettingsEvent?) {
        guard let event else { return }
        switch event {
        case let .showMessage(text):
            showMessage(text)
        case .signedOut:
            onSignedOut()
        }
        viewModel.consumeEvent()
    }

    private func showMessage(_ text: String) {
        messageDismissTask?.cancel()
        withAnimation(.snappy) {
            message = text
        }
        messageDismissTask = Task {
            try? await Task.sleep(for: .seconds(2.5))
            guard !Task.isCancelled else { return }
            withAnimation(.snappy) {
                message = nil
            }
        }
    }

    static var appVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
            ?? "-"
    }
}

private enum Metrics {
    static let topPadding: CGFloat = 48
    static let appToInformationSpacing: CGFloat = 36
    static let loadingHeight: CGFloat = 56
    static let toastBottomPadding: CGFloat = 88
}

#Preview("Member settings") {
    SettingsScreen(
        viewModel: SettingsViewModel(
            initialState: SettingsViewState(
                isLoading: false,
                isLoggedIn: true,
                signatureURL: nil,
                version: "1.0"
            ),
            isAuthenticated: { true }
        )
    )
        .chalkakTheme(.light)
}

#Preview("Guest settings") {
    SettingsScreen(
        viewModel: SettingsViewModel(
            initialState: SettingsViewState(isLoading: false, version: "1.0")
        )
    )
        .chalkakTheme(.light)
}
